test/k8sT/Services.go

```diff
@@ -1269,13 +1272,56 @@ var _ = Describe("K8sServicesTest", func() {
 
 			// With ClusterIP
-			doFragmentedRequest(clientPod, srcPort, serverPort, data.Spec.ClusterIP, data.Spec.Ports[1].Port, true)
+			doFragmentedRequest(clientPod, srcPort, serverPort, data.Spec.ClusterIP, data.Spec.Ports[1].Port, testDS, true)
 
 			// From pod via node IPs
-			doFragmentedRequest(clientPod, srcPort+1, serverPort, k8s1IP, nodePort, hasDNAT)
-			doFragmentedRequest(clientPod, srcPort+2, serverPort, "::ffff:"+k8s1IP, nodePort, hasDNAT)
-			doFragmentedRequest(clientPod, srcPort+3, serverPort, k8s2IP, nodePort, hasDNAT)
-			doFragmentedRequest(clientPod, srcPort+4, serverPort, "::ffff:"+k8s2IP, nodePort, hasDNAT)
+			doFragmentedRequest(clientPod, srcPort+1, serverPort, k8s1IP, nodePort, testDS, hasDNAT)
+			doFragmentedRequest(clientPod, srcPort+2, serverPort, "::ffff:"+k8s1IP, nodePort, testDS, hasDNAT)
+			doFragmentedRequest(clientPod, srcPort+3, serverPort, k8s2IP, nodePort, testDS, hasDNAT)
+			doFragmentedRequest(clientPod, srcPort+4, serverPort, "::ffff:"+k8s2IP, nodePort, testDS, hasDNAT)
+		}
+
+		testIPv4FragmentMetrics := func() {
+			var (
+				data    v1.Service
+				srcPort = 12345
+				hasDNAT = true
+			)
+
+			ciliumPodK8s2, err := kubectl.GetCiliumPodOnNode(helpers.K8s2)
+			ExpectWithOffset(2, err).Should(BeNil(), "Cannot get cilium pod on k8s2")
+
+			// Destination address and port for fragmented datagram
+			// are not DNAT-ed with kube-proxy but without bpf_sock.
+			if helpers.RunsWithKubeProxy() {
+				hasDNAT = kubectl.HasHostReachableServices(ciliumPodK8s2, false, true)
+			}
+
+			clientPodK8s1, _ := kubectl.GetPodOnNodeLabeledWithOffset(helpers.K8s1, testDSClient, 1)
+
+			// Pick test-nodeport-local-k8s2 as service/destination
+			// port, which has a `Local` `externalTrafficPolicy`.
+			//
+			// This will ensure packets hitting k8s2 will not be
+			// fowarded back to the k8s1 node.
+			err = kubectl.Get(helpers.DefaultNamespace, "service test-nodeport-local-k8s2").Unmarshal(&data)
+			ExpectWithOffset(1, err).Should(BeNil(), "Cannot retrieve service")
+			nodePort := data.Spec.Ports[1].NodePort
+			serverPort := data.Spec.Ports[1].TargetPort.IntValue()
+
+			cmd := "cilium bpf metrics list | grep 'Fragmented packet' | grep INGRESS | awk '{print $4}'"
+			res := kubectl.CiliumExecMustSucceed(context.TODO(), ciliumPodK8s2, cmd)
+			fragmentedPacketsBefore, _ := strconv.Atoi(strings.TrimSpace(res.Stdout()))
+
+			fragments := doFragmentedRequest(clientPodK8s1, srcPort+5, serverPort, k8s2IP, nodePort, testDSK8s2, hasDNAT)
+
+			cmd = "cilium bpf metrics list | grep 'Fragmented packet' | grep INGRESS | awk '{print $4}'"
+			res = kubectl.CiliumExecMustSucceed(context.TODO(), ciliumPodK8s2, cmd)
+			fragmentedPacketsAfter, _ := strconv.Atoi(strings.TrimSpace(res.Stdout()))
+
+			Expect(fragmentedPacketsAfter-fragmentedPacketsBefore).Should(Equal(fragments), "Failed to account for IPv4 fragments in BPF metrics")
+
+			// TODO: "Fragmented packet entry update failed" metric
```

Comment:
I'm still looking into this, perhaps I can keep it for a subsequent PR?

I have some Go code to create IP packets:

```go
	ipLayer := layers.IPv4{
		Version: 4,
		Id:      4,
		//Flags:      layers.IPv4MoreFragments,
		//FragOffset: 64,
		Protocol: layers.IPProtocolUDP,
		SrcIP:    net.ParseIP("192.168.1.42"),
		DstIP:    net.ParseIP("192.168.1.1"),
	}
	udpLayer := layers.UDP{
		SrcPort: layers.UDPPort(1234),
		DstPort: layers.UDPPort(1234),
	}
	udpLayer.SetNetworkLayerForChecksum(&ipLayer)

	buf := gopacket.NewSerializeBuffer()
	opts := gopacket.SerializeOptions{
		FixLengths:       true,
		ComputeChecksums: true,
	}

	err := gopacket.SerializeLayers(buf, opts, &ipLayer, &udpLayer, gopacket.Payload("data"))
	if err != nil {
		panic(err)
	}

	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_RAW)
	if err != nil {
		panic(err)
	}
	defer syscall.Close(fd)

	err = syscall.Sendto(fd, buf.Bytes(), 0, &syscall.SockaddrInet4{})
	if err != nil {
		panic(err)
	}
```
but it looks like raw sockets don't play well with fragmentation as the moment I set the MF flag or the frag offset field, I stop seeing the packet on tcpdump :thinking:

Before resorting to iptables (which is something I'd like to avoid) I'll try to figure out if I can somehow make it work with raw sockets.

Reply:
No objection from my side :+1:

Reply:
Thinking again on the map update failure to test, we can maybe create a LRU map frozen on the BPF side prior to starting the Cilium agent, and make sure that it gets reused by the eBPF program. I think that could work. Could be part of the follow-up too.

Reply:
I think it's fine if we don't test that one metric. I doubt we tests the other equivalent metrics in the datapath and, in any case, we have to consider whether it's worth the additional test code complexity.